controllers/nodehealthcheck_controller.go

```diff
@@ -336,41 +372,36 @@ func (r *NodeHealthCheckReconciler) remediate(ctx context.Context, n *v1.Node, n
 
 		// create CR
 		r.Log.Info("node seems unhealthy. Creating an external remediation object",
-			"nodeName", n.Name, "CR name", cr.GetName(), "CR gvk", cr.GroupVersionKind(), "ns", cr.GetNamespace())
+			"nodeName", node.Name, "CR name", cr.GetName(), "CR gvk", cr.GroupVersionKind(), "ns", cr.GetNamespace())
 		if err = r.Client.Create(ctx, cr); err != nil {
 			r.Log.Error(err, "failed to create an external remediation object")
 			return nil, err
 		}
-		r.Recorder.Event(nhc, eventTypeNormal, eventReasonRemediationCreated, fmt.Sprintf("Created remediation object for node %s", n.Name))
+		r.Recorder.Event(nhc, eventTypeNormal, eventReasonRemediationCreated, fmt.Sprintf("Created remediation object for node %s", node.Name))
 		return nil, nil
 	}
 
 	// CR exists
 	isAlert, nextReconcile := r.alertOldRemediationCR(cr)
 	if isAlert {
-		metrics.ObserveNodeHealthCheckOldRemediationCR(n.Name, n.Namespace)
+		metrics.ObserveNodeHealthCheckOldRemediationCR(node.Name, node.Namespace)
 	}
 	return nextReconcile, nil
 }
 
-func (r *NodeHealthCheckReconciler) generateRemediationCR(n *v1.Node, nhc *remediationv1alpha1.NodeHealthCheck) (*unstructured.Unstructured, error) {
-	t, err := r.fetchTemplate(nhc)
-	if err != nil {
-		return nil, err
-	}
-
-	templateSpec, found, err := unstructured.NestedMap(t.Object, "spec", "template")
+func (r *NodeHealthCheckReconciler) generateRemediationCR(n *v1.Node, nhc *remediationv1alpha1.NodeHealthCheck, template *unstructured.Unstructured) (*unstructured.Unstructured, error) {
```

Comment:
Instead of trickling the `template` all the way down, maybe save it on the NodeHealthCheckReconciler.

Reply:
While I agree it's not nice, you can't save anything specific to a single reconcile request in the reconciler, because it potentially can reconcile multiple requests in parallel.
https://pkg.go.dev/github.com/kubernetes-sigs/controller-runtime/pkg/controller?utm_source=godoc#Options

I'm thinking about a passing a new request specific struct around with the template, inflightrmediations and probably some more, but that would be too much for this PR IMHO.